pulsar-broker/src/main/java/org/apache/pulsar/broker/web/PulsarWebResource.java

```diff
@@ -493,13 +494,39 @@ private URI getRedirectionUrl(ClusterData differentClusterData) throws Malformed
             } else {
                 serviceNameResolver.updateServiceUrl(differentClusterData.getServiceUrl());
             }
-            URL webUrl = new URL(serviceNameResolver.resolveHostUri().toString());
+            URL webUrl = getConnectableURL(serviceNameResolver);
             return UriBuilder.fromUri(uri.getRequestUri()).host(webUrl.getHost()).port(webUrl.getPort()).build();
         } catch (PulsarClientException.InvalidServiceURL exception) {
             throw new MalformedURLException(exception.getMessage());
         }
     }
 
+    private URL getConnectableURL(PulsarServiceNameResolver serviceNameResolver) throws MalformedURLException {
```

Comment:
We should make this method pure async. This is called from the metadata callback thread.